In [10]:
import pandas as pd 
import numpy as np 
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

In [3]:
data = pd.read_csv("./Coleccionista.csv")

In [4]:
data.shape

(25, 4)

In [5]:
data.head()

Painting  Price  Bidders  Age
0         1   3470       10   67
1         2   3500        8   56
2         3   3700        7   73
3         4   3860        4   71
4         5   3920       12   99

## Variables independientes

In [6]:
X = data[['Bidders','Age']]
X

Bidders  Age
0        10   67
1         8   56
2         7   73
3         4   71
4        12   99
5        10   87
6        11   78
7         8   83
8        13   90
9        13   98
10        0   91
11        7   93
12        2   97
13        6   79
14        4   83
15        3   71
16        9   89
17        5  103
18        3  106
19        8   93
20        8   88
21        4   96
22        5   94
23        3   88
24        1   84

## Variable dependiente

In [7]:
y = data[['Price']]
y

Price
0    3470
1    3500
2    3700
3    3860
4    3920
5    3900
6    3830
7    3940
8    3880
9    3940
10   4200
11   4060
12   4200
13   4020
14   4190
15   4130
16   4130
17   4370
18   4450
19   4390
20   4380
21   4540
22   4660
23   4710
24   4880

In [12]:
model.intercept_

array([3080.05303581])

In [11]:
#Ajustar el modelo de regresion multiple
model = LinearRegression()
model.fit(X,y) #entrenando el modelo
y_pred = model.predict(X)
#y_pred 

#Metricas estadisticas
mse = mean_squared_error(y, y_pred) #Error de la media cuadratica
r2 = r2_score(y, y_pred) #varianza

In [13]:
model.intercept_

array([3080.05303581])

In [14]:
print('Coeficientes: ', model.coef_)
print('Error cuadratico de la media: ', mse)
print('Varianza: ', r2)

Coeficientes:  [[-54.18854049  16.2891028 ]]
Error cuadratico de la media:  43396.435062885874
Varianza:  0.647801948911782


## ¿Le sorpende que exista una relacion inversa entre el numero de postores y el precio de la pintura?
Si es sorprendete porque cuando uno ve una subasta nota que sube el precio cuanto mas postores existe por el hecho de la competencia que existe dentro del evento, asi que si llega a ser sorprendente el resultado


# Incluir una nueva variable y ver si es significativa o no


In [15]:
import random

In [16]:
random.seed(42)
firma = [random.randint(0,1) for _ in range(25)]

data['Firma'] = firma

In [17]:
print(data[['Firma']].head())

   Firma
0      0
1      0
2      1
3      0
4      0


In [18]:
print('Coeficientes: ', model.coef_)
print('Error cuadratico de la media: ', mse)
print('Varianza: ', r2)

Coeficientes:  [[-54.18854049  16.2891028 ]]
Error cuadratico de la media:  43396.435062885874
Varianza:  0.647801948911782


In [19]:
X = data[['Bidders','Age','Firma']]
X

Bidders  Age  Firma
0        10   67      0
1         8   56      0
2         7   73      1
3         4   71      0
4        12   99      0
5        10   87      0
6        11   78      0
7         8   83      0
8        13   90      1
9        13   98      0
10        0   91      0
11        7   93      0
12        2   97      0
13        6   79      0
14        4   83      0
15        3   71      0
16        9   89      1
17        5  103      0
18        3  106      1
19        8   93      1
20        8   88      0
21        4   96      0
22        5   94      1
23        3   88      1
24        1   84      1

In [20]:
model.intercept_

array([3080.05303581])

In [21]:
model = LinearRegression()
model.fit(X,y) #entrenando el modelo
y_pred = model.predict(X)
#y_pred 

#Metricas estadisticas
mse = mean_squared_error(y, y_pred) #Error de la media cuadratica
r2 = r2_score(y, y_pred) #varianza

In [22]:
print('Coeficientes: ', model.coef_)
print('Error cuadratico de la media: ', mse)
print('Varianza: ', r2)

Coeficientes:  [[-52.20201798  14.6379248  218.12979959]]
Error cuadratico de la media:  33488.739726566404
Varianza:  0.7282111111660303


Si es significativa la variable Firma ya que este hizo que aumente la varianza a 0.73 de lo que estaba en 0.65, por ende, nuestro modelo mejoró.